### PR TITLE
Complex volume visual

### PR DESCRIPTION
**Pull request by tlambert03**
Follow up to #1999, which would add support for complex valued data to volumes as well.  WIP... includes #1999.  leaving as draft for now until that one merges.

### CODE REVIEW

**Comment by roflmaostc** [comment]
Is there some progress :)?
It's really one feature which would make Napari really great!

**Comment by roflmaostc** [comment]
Just wanted to follow-up :smile:

**Comment by tlambert03** [comment]
👋 
I'm afraid to say I'm not really working on napari anymore so someone will need to pick this up

**Comment by roflmaostc** [comment]
Ah, that's sad to hear.

How much of the work is missing to get this done here? Is it just for volumes missing?

**Comment by tlambert03** [comment]
I can't remember. I'm not sure it's necessarily work that needs to be done as much as agreement on APIs, etc

**Comment by tlambert03** [comment]
I'll resolve the conflicts today

**Comment by tlambert03** [comment]
conflicts resolved!

File: examples/scene/complex_volume.py [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+cam = scene.cameras.TurntableCamera(parent=view.scene, fov=60, name="Turntable")
+# cam = scene.cameras.ArcballCamera(parent=view.scene, fov=60, name='Arcball')
+view.camera = cam
+
+# Create the image
+img_data = complex_ramp()
+
+image = scene.visuals.ComplexVolume(
+    img_data.real, parent=view.scene, complex_mode="magnitude"
+)
```
Comment (djhoese) [inline examples/scene/complex_volume.py:39]: Should this now pass `img_data` and not just the `.real` part?

File: examples/scene/complex_volume.py [hunk 2]
```diff
@@ -0,0 +1,48 @@
+# -*- coding: utf-8 -*-
+# vispy: gallery 30
+# -----------------------------------------------------------------------------
+# Copyright (c) Vispy Development Team. All Rights Reserved.
+# Distributed under the (new) BSD License. See LICENSE.txt for more info.
+# -----------------------------------------------------------------------------
+"""
+Simple use of SceneCanvas to display an Image.
+"""
```
Comment (djhoese) [inline examples/scene/complex_volume.py:9]:
```suggestion
View volume of complex numbers
==============================
"""
```

**Review by djhoese — CHANGES_REQUESTED** [review]

**Comment by tlambert03** [comment]
@brisvag, it would be great if you could pick this up.  I don't really remember the details and am no longer invested in the outcome

**Comment by brisvag** [comment]
@djhoese what do you think, should this all be ported to `VolumeVisual` to make it handle anything, or should `ComplexVolumeVisual` just handle complex data and that's it?

**Comment by djhoese** [comment]
I'm not sure how I feel and I'm sure that I've flip flopped on my opinion multiple times over the life of these volume-based PRs. I feel like this feature and its implementation is right on the edge of being too much to add on to the existing `VolumeVisual` but arguably too little for its own Visual. I barely understand the VolumeVisual as it is so this is even harder for me to have an opinion about that I feel comfortable with. I think if someone can implement this functionality in the VolumeVisual and make it understandable/readable then I'm OK with that. In general VisPy's Visuals have too many options as it is so it is hard to justify doing something against that (admittedly bad) design practice or rather it is hard to say "oh but that would be too many kwargs in VolumeVisual" when VisPy does this so much already.

**Comment by brisvag** [comment]
Now that I look around, I see that we already have the `ComplexImageVisual` which basically follows the same api as this... So maybe for consistency, and given that the `ComplexImage` is already out for a while, it's better to follow the same line.

**Comment by brisvag** [comment]
Created a new PR at #2539, closing here.

**Pull request closed without merging**